### PR TITLE
cmake: add USE_SYSTEM_{LIBS,CPUINFO,SLEEF} options (#14699) 

### PR DESCRIPTION
**Pull request by cdluminate**
@ezyang

### CODE REVIEW

**Comment by cdluminate** [comment]
I've already packaged cpuinfo for Debian. This looks like a good start.

File: aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt [hunk 1]
```diff
@@ -60,6 +60,7 @@ if(NOT DEFINED CLOG_SOURCE_DIR)
   set(CLOG_SOURCE_DIR "${PROJECT_SOURCE_DIR}/deps/clog")
 endif()
 
+if(NOT USE_SYSTEM_CPUINFO)
```
Comment (peterjc123) [inline aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:63]: nit: indent

File: CMakeLists.txt [hunk 2]
```diff
@@ -207,6 +207,13 @@ cmake_dependent_option(
 option(USE_TBB "Use TBB" OFF)
 option(ONNX_ML "Enable traditional ONNX ML API." ON)
 
+# Linux distributions do not want too many embedded sources
+option(USE_SYSTEM_LIBS "Use system-provided libraries." OFF)
+option(USE_SYSTEM_CPUINFO "Use system-provided cpuinfo." OFF)
+if(USE_SYSTEM_LIBS)
+	set(USE_SYSTEM_CPUINFO ON)
```
Comment (peterjc123) [inline CMakeLists.txt:214]: nit: 2 spaces

File: aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt [hunk 3]
```diff
@@ -302,7 +304,12 @@ endif()
 target_link_libraries(pytorch_qnnpack PUBLIC clog)
 
 # ---[ Configure cpuinfo
-if(NOT TARGET cpuinfo)
+if(NOT TARGET cpuinfo AND USE_SYSTEM_CPUINFO)
+  add_library(cpuinfo SHARED IMPORTED)
+  find_library(LIBCPUINFO cpuinfo)
```
Comment (peterjc123) [inline aten/src/ATen/native/quantized/cpu/qnnpack/CMakeLists.txt:309]: Rename to `CPUINFO_LIBRARY`?

File: cmake/Dependencies.cmake [hunk 4]
```diff
@@ -310,7 +310,12 @@ if(USE_XNNPACK)
 endif()
 
 # ---[ Caffe2 uses cpuinfo library in the thread pool
-if(NOT TARGET cpuinfo)
+if(NOT TARGET cpuinfo AND USE_SYSTEM_CPUINFO)
+  add_library(cpuinfo SHARED IMPORTED)
+  find_library(LIBCPUINFO cpuinfo)
```
Comment (peterjc123) [inline cmake/Dependencies.cmake:315]: Rename to `CPUINFO_LIBRARY`?

**Comment by cdluminate** [comment]
@peterjc123 fixed and force-pushed.

File: CMakeLists.txt [hunk 5]
```diff
@@ -207,6 +207,19 @@ cmake_dependent_option(
...
+option(USE_SYSTEM_CPUINFO "Use system-provided cpuinfo." OFF)
+option(USE_SYSTEM_SLEEF "Use system-provided sleef." OFF)
+if(DEFINED ENV{USE_SYSTEM_LIBS})
+  set(USE_SYSTEM_LIBS "$ENV{USE_SYSTEM_LIBS}")
+endif()
+if(USE_SYSTEM_LIBS)
+  set(USE_SYSTEM_CPUINFO ON)
+  set(USE_SYSTEM_SLEEF ON)
+  set(BUILD_CUSTOM_PROTOBUF OFF)
+endif()
```
Comment (ezyang) [inline CMakeLists.txt:221]: @xuhdev would you mind checking this and saying if `cmake_dependent_option` would be appropriate here?

File: CMakeLists.txt [hunk 6]
```diff
@@ -207,6 +207,19 @@ cmake_dependent_option(
 option(USE_TBB "Use TBB" OFF)
 option(ONNX_ML "Enable traditional ONNX ML API." ON)
 
+# Linux distributions do not want too many embedded sources
+option(USE_SYSTEM_LIBS "Use system-provided libraries." OFF)
+option(USE_SYSTEM_CPUINFO "Use system-provided cpuinfo." OFF)
+option(USE_SYSTEM_SLEEF "Use system-provided sleef." OFF)
+if(DEFINED ENV{USE_SYSTEM_LIBS})
```
Comment (ezyang) [inline CMakeLists.txt:214]: I don't think you need the explicit env var check here; setup.py will handle converting environment variables into cmake flags.

**Comment by ezyang** [comment]
This LGTM. Just some bits at the top level. Can we put some description about how to test these flags? It probably isn't good ROI to add these to the CI, but at least there should be manual testing instructions.

File: CMakeLists.txt (hunk 6 above)
Comment (cdluminate) [reply CMakeLists.txt:214]: This env check is convenient for my local testing. I can try to add replacement code to setup.py for the next PR.

**Comment by cdluminate** [comment]
> This LGTM. Just some bits at the top level. Can we put some description about how to test these flags? It probably isn't good ROI to add these to the CI, but at least there should be manual testing instructions.

This is my testing procedure:
1. git clone <pytorch>, then init and update the git submodules, recursively
2. remove 3rdparty libraries that have already been packaged by debian, e.g. sleef (`libsleef-dev`) and cpuinfo (`libcpuinfo-dev`).
3. `export USE_SYSTEM_LIBS=ON`
4. `python3 setup.py install`

Testing environment: Debian unstable, with corresponding build dependencies installed.

My ultimate goal is to make pytorch buildable with an (almost) empty `third_party` directory (.e.g no any submodules has been pulled). In that case pytorch would be ready to be uploaded to Debian. Apart from that, I'm still packaging the build dependencies for Debian. Subsequent PRs will be submitted in the future.

**Comment by malfet** [comment]
@cdluminate can you please split this into two PRs (or update PR description to say that multiple `USE_SYSTEM_...` flags are added)?

File: CMakeLists.txt (hunk 5 above)
Comment (xuhdev) [reply CMakeLists.txt:221]: If `USE_SYSTEM_LIBS` is `OFF`, does it mean the three options won't be available? If so, I think `cmake_dependent_option` is appropriate.

File: CMakeLists.txt (hunk 6 above)
Comment (xuhdev) [reply CMakeLists.txt:214]: @cdluminate You don't need to add anything to `setup.py`. Simply remove this `if` block and it'll be automatically passed to CMake in https://github.com/pytorch/pytorch/blob/438aed63a143103180eeff3e8991413e74569563/tools/setup_helpers/cmake.py#L251-L261

File: CMakeLists.txt (hunk 5 above)
Comment (cdluminate) [reply CMakeLists.txt:221]: USE_SYSTEM_LIBS is a shortcut variable to toggle all the USE_SYSTEM_* variables on. Users can still toggle e.g. USE_SYSTEM_SLEEF without turning USE_SYSTEM_LIBS on. I'll write it to the comments in the next update.

**Comment by cdluminate** [comment]
I have resolved the issues and added some comments to make things clear. Commits have been squashed into one, with the commit message begin rewritten.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**